File: Quick-Sort.py
# Create a function to find partition position
# array[ ] ---> Array to be sorted
# low ---> Starting index
# high ---> Ending index
def partition_pos(array, low, high):
    
    # Set the rightmost element as pivot
    pivot = array[high]
    
    # Index of smaller element
    i = low - 1
    
    # Traverse through all the element and compare each element with pivot
    for j in range(low, high):
        
         # If current element is smaller than or equal to pivot
         if array[j] <= pivot:
             
             # increment the index of smaller element
             i += 1
             
             # swapping element at i with element at j
             array[i], array[j] = array[j], array[i]
             
    # swap the pivot element with the greater element specified by i
    array[i+1], array[high] = array[high], array[i+1]
    
    # return the position from where partition is done
    return i + 1

# Create a main function that implements quicksort    
def quick_sort(array, low, high):
    
    # Find pivot element such that element smaller than pivot are on the left element greater than pivot are on the right    
    if low < high:
        
        # pi is partitioning index, arr[pi] is now at right place
        pi = partition_pos(array, low, high)
        
        # recursive call on the left of pivot
        quick_sort(array, low, pi - 1)
        
        # recursive call on the right of pivot
        quick_sort(array, pi+1, high)
     
array = [1, 9, 45, 23, 12, 82, 67]
n = len(array)
quick_sort(array, 0, n-1)
print(array)

File: Quick-Sort.c
// Quick sort in C
#include <stdio.h>

// function to find the partition position
int partition(int array[], int low, int high)
{
    int pivot, leftIndex, i, temp;
    
    // select the rightmost element as pivot
    pivot = array[high];
    
    // pointer for greater element
    leftIndex = -1;
    
    // traverse each element of the array 
    // compare them with the pivot
    for (i = 0; i < high; i++)
    {
        if (array[i] <= pivot)
        {
            // if element smaller than pivot is found 
            // swap it with the greater element pointed by leftIndex
            leftIndex++;
            
            // swap element at leftIndex with element at i
            temp = array[leftIndex];
            array[leftIndex] = array[i];
            array[i] = temp;
        }
    }
 
    // swap the pivot element with the greater element at leftIndex  
    temp = array[leftIndex+1];
    array[leftIndex+1] = array[high];
    array[high] = temp;
    
    // return the partition point
    return (leftIndex + 1);   
}

void quickSort(int array[], int low, int high)
{
    int index;
    if (low < high)
    {
        // find the pivot element such that 
        // elements smaller than pivot are on left of pivot 
        // elements greater than pivot are on right of
        index = partition(array, low, high);
        
        // recursive call on the left of pivot
        quickSort(array, low, index-1);
        
        // recursive call on the right of pivot
        quickSort(array, index+1, high);
    }
}

// function to print array elements 
void printArray(int array[], int size) 
{ 
    for (int i = 0; i < size; ++i)
    { 
        printf("%d ", array[i]); 
    } 
    printf("\n"); 
} 
// main function 
int main() 
{ 
    int data[] = {8, 7, 2, 1, 0, 9, 6}; 
    int n = sizeof(data) / sizeof(data[0]); 
    printf("Unsorted Array\n"); 
    printArray(data, n); 
    
    // perform quicksort on data 
    quickSort(data, 0, n - 1); 
    printf("Sorted array in ascending order: \n"); 
    printArray(data, n); 
}
